## Final Project Submission

Please fill out:
* Student name: VALENTINE MARANGALA KWEYU
* Student pace: part time
* Scheduled project review date/time: 
* Instructor name: MWIKALI
* Blog post URL: 


## PROJECT OVERVIEW
For this project, we will use data cleaning, imputation, analysis, and visualization to generate insights for a business stakeholder interested in purchasing and operating airplanes for commercial & private enterprises.

### Business Understanding
The company is expanding in to new industries to diversify its portfolio. Specifically, the company is interested in purchasing and operating aircrafts, but is clueless about the potential risks of aircraft. This project wil be used as a criteria to determine which aircraft has the lowest risk of accidents, for the company to start this new business endeavor. The findings will be converted to actionable insights to aid head of the new aviation division decide which aircraft to purchase.

### Goals
* To identify which aircraft models have the lowest accidents.
* to identify how external factors impact the risk.
* To assess whether specific features in the aircraft is associated with lower risks. eg engine numbers.

### Data Understanding

#### Data
The data was obtained from Kaggle datasets (Aviation Accident Database & Synopses, up to 2023).The National Transportation Safety Board(NTSB) aviation accident **[database](https://www.kaggle.com/datasets/khsamaha/aviation-accident-database-synopses/data)** contains information from 1962 to 2023 about civil aviation accidents and selected incidents within the United States, its territories and possessions, and in international waters . This dataset will form the basis of trying to establish the best aircraft to adapt.

#### Loading Data
The dataset is in **.csv** format.  Therefore we shall use **pandas** with alias **pd** to read the CSV file.



In [120]:
# Load the data
# Our data has mixed types in Columns(6,7,28)
# Therfore we specify the dtype option with the set function.
import pandas as pd 
df = pd.read_csv('data\Aviation_Data.csv', low_memory = False) 


#### Data exploration
We proceed on by performing preliminary data exploring on our dataframe, which is denoted as  **df**. This process aids us to understand our data properties, types and structures.

Here’s a structured breakdown of your dataset’s initial inspection, highlighting key issues.

In [121]:
# For initial Checks:
# Use the df.head method inspects the first 5 rows in our dataframe.
df.head()           

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [122]:
# df.info() summarizes the dataset's structure.

df.info()

print(f"Count of null values in column (Model) :", df['Model'].isnull().sum()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

The output gives us some useful insights about our dataframe:
1. **Size**: We have 90348 entries that contains missing values. (eg. **Model** column that has 1551 nulls values out of 90348 rows.)

2. **Data types**: The dataset has  31 columns. Where 5 columns are of type **float** and 26 columns of type **object**.

3. **Memory**: 21.4+ MB that is manageable; no need to downcast yet.




In [123]:
# Descriptive statistics
# .describe() method

df.describe()

Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

The code shows a focused analysis of the numeric columns in the dataset (**Number.of.Engines, Total.Fatal.Injuries, Total.Serious.Injuries, Total.Minor.Injuries, Total.Uninjured**). 

If we look at some of the columns , we realize that :

1. **Number.of.Engines:**

         Mean: 1.15 meaning most aircrafts have single engines.

         Max: The maximum engines inside an aircraft is 8.(mostly large commercial jets)

    This indicates that 75% of aircrafts have 1 engine which correlates with the general aviation dominance in the dataset.



2. **Injury Statistics:**

    -**Fatality Rate**:

         Mean = 0.65 fatalities per accident

         Max = 349 (mid-air collisions like 9/11 event)

    Key Finding: 75% of accidents have 0 fatalities (median=0), but extreme outliers exist.



3. **Survivability(Total.Uninjured):**

        Uninjured Mean: 5.3 people unharmed per accident

        Max: 699 uninjured (likely a survivable commercial jet incident)


In [124]:
# Missing Values
# df.isnull() method is used to count missing values per column.

missing_percent = df.isnull().mean() * 100          # converts to percentage
print(missing_percent.sort_values(ascending=False)) # returns columns in descending order

Schedule                  86.073848
Air.carrier               81.573471
FAR.Description           64.555939
Aircraft.Category         64.263736
Longitude                 61.954886
Latitude                  61.944924
Airport.Code              44.382831
Airport.Name              41.570372
Broad.phase.of.flight     31.681941
Publication.Date          18.471909
Total.Serious.Injuries    15.461327
Total.Minor.Injuries      14.822686
Total.Fatal.Injuries      14.233851
Engine.Type                9.447913
Report.Status              8.677558
Purpose.of.flight          8.468367
Number.of.Engines          8.348829
Total.Uninjured            8.158454
Weather.Condition          6.586753
Aircraft.damage            5.150086
Registration.Number        3.072564
Injury.Severity            2.721698
Country                    1.865011
Amateur.Built              1.727764
Model                      1.716695
Make                       1.684597
Location                   1.672422
Event.Date                 1

According to the output above we can categorize our data into 4 Comprehensive Missing Data Analysis groups:

1. **Critical Columns (High Priority consider dropping rows) :**

         Model             - 1551 missing values

         Make              - 1522 missing values

         Injury.Severity   - 2459 missing values

         Engine.Type       - 8536 missing values


2. **Moderate-Priority Columns (Consider adding/replacing values) :**

         Weather.Condition - 5,951 missing values

         Number.of.Engines - 7543 missing values

         Total.Fatal.Injuries - 12860 missing values

3.  **Low-Priority Columns (Consider Dropping) :**

         Latitude/Longitude - 55966 missing values

         FAR.Description - 58,325 missing values
        
         Schedule - 77,766 missing values

4. **Special Cases :**
        
        Aircraft.Category - 58,061 missing values (If needeed we shall  impute "UNKNOWN" or otherwise drop)

        Broad.phase.of.flight - 28,624 missing values (Impute   "UNKNOWN" )


        


### Data Preparation
In this section we shall be effectively handling the missing data and the following actions will be employed:

- Dropping Irrelevant Columns (*with 45% and above missing values*)

- Handling missing data (*critical columns, numeric columns, categorical columns*)

- Fixing Data Types

In [125]:
#  Drop Irrelevant Columns

cols_to_drop = ['FAR.Description','Schedule', 'Air.carrier', 'Aircraft.Category', 'Latitude', 'Longitude', 'Airport.Code', 'Airport.Name']
df_clean = df.drop(columns=cols_to_drop)
missing_percent[missing_percent > 40]

Latitude             61.944924
Longitude            61.954886
Airport.Code         44.382831
Airport.Name         41.570372
Aircraft.Category    64.263736
FAR.Description      64.555939
Schedule             86.073848
Air.carrier          81.573471
dtype: float64

In [126]:
# Handling Missing Data

## Critical Columns - we are removing rows with missing data in critical columns.
df_clean = df_clean.dropna(subset=['Model', 'Make', 'Injury.Severity', 'Engine.Type'])


## Numeric Columns (e.g., Injuries) - assume no injury if null
injury_cols = ['Total.Fatal.Injuries', 'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured']
df_clean[injury_cols] = df_clean[injury_cols].fillna(0) 


## Categorical Columns: - assign 'UNKNOWN' to the missing values
cat_cols = ['Weather.Condition', 'Broad.phase.of.flight', 'Engine.Type']
df_clean[cat_cols] = df_clean[cat_cols].fillna('UNKNOWN')


In [127]:
# Fix Data Types

## Convert Event.Date column to datetime:
df_clean['Event.Date'] = pd.to_datetime(df_clean['Event.Date'], errors='coerce')
df_clean


## Convert Number.of.Engines to integer (after imputing):
df_clean['Number.of.Engines'] = df_clean['Number.of.Engines'].fillna(1).astype(int)

